nx/lib/nx.ex

```diff
@@ -1228,6 +1217,199 @@ defmodule Nx do
     raise ArgumentError, "diag/2 expects tensor of rank 1 or 2, got: #{inspect(tensor)}"
   end
 
+  @doc """
+  Extracts the diagonal of a 2D tensor.
+
+  Converse of `make_diagonal/2`.
+
+  ## Examples
+
+  Given a 2D tensor without offset:
+
+      iex> Nx.take_diagonal(Nx.tensor([
+      ...> [0, 1, 2],
+      ...> [3, 4, 5],
+      ...> [6, 7, 8]
+      ...> ]))
+      #Nx.Tensor<
+        s64[3]
+        [0, 4, 8]
+      >
+
+  And if given a 2D tensor along with an offset:
+
+      iex> Nx.take_diagonal(Nx.iota({3, 3}), offset: 1)
+      #Nx.Tensor<
+        s64[2]
+        [1, 5]
+      >
+
+      iex> Nx.take_diagonal(Nx.iota({3, 3}), offset: -1)
+      #Nx.Tensor<
+        s64[2]
+        [3, 7]
+      >
+
+  ## Options
+
+    * `:offset` - offset used for extracting the diagonal.
+      Use offset > 0 for diagonals above the main diagonal,
+      and offset < 0 for diagonals below the main diagonal.
+
+  ## Error cases
+
+      iex> Nx.take_diagonal(Nx.tensor([0, 1, 2]))
+      ** (ArgumentError) take_diagonal/2 expects tensor of rank 2, got:
+      #Nx.Tensor<
+        s64[3]
+        [0, 1, 2]
+      >
+
+      iex> Nx.take_diagonal(Nx.iota({3, 3}), offset: 3)
+      ** (ArgumentError) offset must be less than breadth when positive, got: 3
+
+      iex> Nx.take_diagonal(Nx.iota({3, 3}), offset: -4)
+      ** (ArgumentError) absolute value of offset must be less than length when negative, got: -4
+  """
+  @doc type: :creation
+  def take_diagonal(tensor, opts \\ []) do
+    opts = keyword!(opts, [:offset])
+
+    shape = Nx.Shape.take_diagonal(tensor)
+    offset = opts[:offset] || 0
+
+    Nx.Shape.validate_diag_offset!(shape, offset)
+
+    Nx.gather(tensor, diag_indices(shape, offset))
+  end
+
+  @doc """
+  Creates a diagonal tensor from a 1D tensor.
+
+  Converse of `take_diagonal/2`.
+
+  If given an offset, the diagonal tensor will be a square matrix
+  with length `size(tensor)` + `Kernel.abs(:offset)`, with the 1D
+  tensor as the diagonal of offset `:offset`.
```

Comment:
I feel this description is can be a bit unclear. Any suggestions?

Reply:
```suggestion
  If given an offset, the resulting tensor will have the corresponding diagonal
  filled with the input tensor's values. The main diagonal as `offset = 0`, whereas
  diagonals above it can be filled with `offset = 1`, `offset = 2` and so on, while the
  diagonals below it can be filled with `offset = -1`, `offset = -2` and so on.
```

Reply:
WDYT about using a less "mathematical" language like this?

Reply:
```suggestion
  The returned tensor will be a square matrix of dimensions equal
  to the size of the tensor. If an offset is given, the absolute value
  of the offset is added to the matrix dimensions sizes.
```

?

Reply:
@Benjamin-Philip please see the suggestion above. :)